// botbuilder-dialogs/lib/dialog.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
var DialogReason;
(function (DialogReason) {
    /**
     * A dialog is being started through a call to `DialogContext.beginDialog()`.
     */
    DialogReason["beginCalled"] = "beginCalled";
    /**
     * A dialog is being continued through a call to `DialogContext.continueDialog()`.
     */
    DialogReason["continueCalled"] = "continueCalled";
    /**
     * A dialog ended normally through a call to `DialogContext.endDialog()`.
     */
    DialogReason["endCalled"] = "endCalled";
    /**
     * A dialog is ending because its being replaced through a call to `DialogContext.replaceDialog()`.
     */
    DialogReason["replaceCalled"] = "replaceCalled";
    /**
     * A dialog was cancelled as part of a call to `DialogContext.cancelAllDialogs()`.
     */
    DialogReason["cancelCalled"] = "cancelCalled";
    /**
     * A step was advanced through a call to `WaterfallStepContext.next()`.
     */
    DialogReason["nextCalled"] = "nextCalled";
})(DialogReason = exports.DialogReason || (exports.DialogReason = {}));
var DialogTurnStatus;
(function (DialogTurnStatus) {
    /**
     * Indicates that there is currently nothing on the dialog stack.
     */
    DialogTurnStatus["empty"] = "empty";
    /**
     * Indicates that the dialog on top is waiting for a response from the user.
     */
    DialogTurnStatus["waiting"] = "waiting";
    /**
     * Indicates that the dialog completed successfully, the result is available, and the stack is empty.
     */
    DialogTurnStatus["complete"] = "complete";
    /**
     * Indicates that the dialog was cancelled and the stack is empty.
     */
    DialogTurnStatus["cancelled"] = "cancelled";
})(DialogTurnStatus = exports.DialogTurnStatus || (exports.DialogTurnStatus = {}));
/**
 * Base class for all dialogs.
 */
class Dialog {
    constructor(dialogId) {
        this.id = dialogId;
    }
    /**
     * (Optional) method called when an instance of the dialog is the active dialog and the user
     * replies with a new activity. The dialog will generally continue to receive the users replies
     * until it calls `DialogContext.endDialog()`, `DialogContext.beginDialog()`, or `DialogContext.prompt()`.
     *
     * If this method is NOT implemented then the dialog will be automatically ended when the user
     * replies.
     * @param dc The dialog context for the current turn of conversation.
     */
    continueDialog(dc) {
        return __awaiter(this, void 0, void 0, function* () {
            // By default just end the current dialog.
            return dc.endDialog();
        });
    }
    /**
     * (Optional) method called when an instance of the dialog is being returned to from another
     * dialog that was started by the current instance using `DialogContext.beginDialog()` or
     * `DialogContext.prompt()`.
     *
     * If this method is NOT implemented then the dialog will be automatically ended with a call
     * to `DialogContext.endDialog()`. Any result passed from the called dialog will be passed to the
     * active dialogs parent.
     * @param dc The dialog context for the current turn of conversation.
     * @param reason The reason the dialog is being resumed. This will typically be a value of `DialogReason.endCalled`.
     * @param result (Optional) value returned from the dialog that was called. The type of the value returned is dependant on the dialog that was called.
     */
    resumeDialog(dc, reason, result) {
        return __awaiter(this, void 0, void 0, function* () {
            // By default just end the current dialog and return result to parent.
            return dc.endDialog(result);
        });
    }
    /**
     * (Optional) method called when the dialog has been requested to re-prompt the user for input.
     * @param context Context for the current turn of conversation.
     * @param instance The instance of the current dialog.
     */
    repromptDialog(context, instance) {
        return __awaiter(this, void 0, void 0, function* () {
            // No-op by default
        });
    }
    /**
     * (Optional) method called when the dialog is ending.
     * @param context Context for the current turn of conversation.
     * @param instance The instance of the current dialog.
     * @param reason The reason the dialog is ending.
     */
    endDialog(context, instance, reason) {
        return __awaiter(this, void 0, void 0, function* () {
            // No-op by default
        });
    }
}
/**
 * Signals the end of a turn by a dialog method or waterfall/sequence step.
 */
Dialog.EndOfTurn = { status: DialogTurnStatus.waiting };
exports.Dialog = Dialog;
