// botbuilder-dialogs/lib/waterfallStepContext.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @module botbuilder-dialogs
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
const dialogContext_1 = require("./dialogContext");
class WaterfallStepContext extends dialogContext_1.DialogContext {
    constructor(dc, info) {
        super(dc.dialogs, dc.context, { dialogStack: dc.stack });
        this._info = info;
    }
    get index() {
        return this._info.index;
    }
    get options() {
        return this._info.options;
    }
    get reason() {
        return this._info.reason;
    }
    get result() {
        return this._info.result;
    }
    get values() {
        return this._info.values;
    }
    next(result) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this._info.onNext(result);
        });
    }
}
exports.WaterfallStepContext = WaterfallStepContext;
